Added opentracing jaeger middleware

sam-burrell opened the pull request:
<!-- 
Please use this template when submitting a new feature request with as much details as possible. Not doing so may result in the issue not being addressed in a timely manner or eventually closed.
-->
# Added opentracing jaeger middleware

## Summary 

Discussion here: https://groups.google.com/forum/#!topic/louketo/Dach4U_Iza8

**Proposal**
Introduce a flag to enable opentracing, under this flag:
Write some middleware to add extraction and injection of traces with some basic debug tags
Initialise a jaeger client to report spans


## Type

[] Bug fix
[] Feature request
[x] Enhancement
[] Docs

## Why?

Keycloak itself uses wildfly. Wildfly introduced support for open tracing https://wildfly.org/news/tags/wildfly/page/2/. There is talk of introducing support to Keycloak
It would be good see the auth flow through opentracing for debugging.  
For us, we would like to integrate the gatekeeper into our auth flow for debugging.

## Requirements

Configured using the standard jaeger client env variables

## How to try it?

Use jaeger-all-in-one https://www.jaegertracing.io/docs/1.18/getting-started/ to test locally.

## Documentation

Can write this as needed.

## Additional Information

I can't see a clean way the proxy closes so right now this is not cleanly closing the jaeger client. Is cleanly shutting down also something worth thinking about?

## Checklist:

Happy to help with anything the further this. 

- [x] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.

asomervell commented:
Why does this introduce so many new dependencies? @sam-burrell it's security software, does it need all that?

Limess committed:
cb/sm: add option to disable csrf; encode state parameter with redirect URL in json

Commit:
sm/cb - enable CSRF check by default

Commit:
sm/cb - update readme

Limess committed:
Add tests for callback handler

evilr00t committed:
print config and RedirectionURL

coro committed:
Prevent privilege escalations through invalid JWTs

Louketo has a bug revolving around this PR: https://github.com/coreos/go-oidc/pull/97
essentially, the JWT verification function first checks the claims of the JWT are valid (including expiry times), then checks the signature of the token is valid.
Louketo checks the response of this and returns a custom error if the token has expired to allow the auth middleware to use a refresh token to regenerate an access token.
Unfortunately, the access token it generates uses the user context for the incoming request. This user context is parsed from the access token JWT but is not verified, meaning it is blindly trusted.
Louketo will then regenerate a correct accessToken from Keycloak, which will be valid and have the correct claim/scopes as generated in Keycloak. I think the issue is that for the rest of the request, it has this new token in the context but still has the leftover rubbish user context parsed from the fake user we injected earlier.

coro committed:
Set the user identity in the scope explicitly

Commit:
Merge pull request #1 from signal-ai/cve-jwt-invalid

Prevent privilege escalations through invalid JWTs

coro committed:
Always reject manipulated JWTs with 401

This results in 401s even for expired tokens
